pkg/controller/genericwebhookconfiguration/certhandler.go

```diff
@@ -0,0 +1,324 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package genericwebhookconfiguration
+
+import (
+	"fmt"
+	"strings"
+
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/kubernetes-sigs/kubebuilder/pkg/webhook/certprovisioner"
+	"net/url"
+)
+
+var (
+	// Use an annotation in the following format:
+	// secret.certprovisioner.kubernetes.io/<webhook-name>: <secret-namespace>/<secret-name>
+	// the webhook cert manager library will provision the certificate for the webhook by
+	// storing it in the specified secret.
+	SecretCertInjectionAnnotationKeyPrefix = "secret.certprovisioner.kubernetes.io/"
+
+	// Use an annotation in the following format:
+	// local.certprovisioner.kubernetes.io/<webhook-name>: path/to/certs/
+	// the webhook cert manager library will provision the certificate for the webhook by
+	// storing it under the specified path.
+	// format: local.certprovisioner.kubernetes.io/webhookName: path/to/certs/
+	LocalCertInjectionAnnotationKeyPrefix = "local.certprovisioner.kubernetes.io/"
+
+	CACertName     = "ca-cert.pem"
+	ServerKeyName  = "key.pem"
+	ServerCertName = "cert.pem"
+)
+
+// CertsHandlerFactory builds CertsHandler.
+type CertsHandlerFactory interface {
+	// New instantiates the CertHandler with a webhookConfiguration object.
+	New(webhookConfiguration runtime.Object) (CertsHandler, error)
+}
+
+// CertsHandler provides methods for handling certificates for webhook.
+type CertsHandler interface {
+	// Skip returns if the webhook should be skipped.
+	Skip(webhookName string) bool
+	// Read reads a wehbook name and returns the certs for it.
+	Read(webhookName string) (*certprovisioner.Certs, error)
+	// Write writes the certs and return the certs it wrote.
+	Write(webhookName string) (*certprovisioner.Certs, error)
+}
+
+type SecretCertsReadWriterFactory struct {
+	KubernetesClientSet kubernetes.Interface
+	KubernetesInformers informers.SharedInformerFactory
```

Comment:
The above two will be replaced after we migrate to controller-runtime.

Reply:
This may happen.